nupic/research/frameworks/dynamic_sparse/models/dsnn.py

```diff
@@ -39,39 +31,48 @@ class DSNNHeb(SparseModel):
     def setup(self):
         super().setup()
 
-        # Set some attributes to be a list corresponding to self.dynamic_sparse_modules.
-        for attr in ["hebbian_prune_perc", "weight_prune_perc"]:
-            self._make_attr_iterable(attr, counterpart=self.sparse_modules)
-        for h_perc, m in zip(self.hebbian_prune_perc, self.sparse_modules):
-            if h_perc is not None:
-                assert hasattr(m, "coactivations"), """
-                Expected {} module to have associated coactivations for Hebbian pruning.
-                """.format(m.__class__.__name__)
-
-        self.added_synapses = [None for m in self.masks]
-        self.last_gradients = [None for m in self.masks]
-
         # add specific defaults
         new_defaults = dict(
             pruning_active=True,
             pruning_early_stop=None,
             hebbian_prune_perc=None,
             weight_prune_perc=None,
             hebbian_grow=True,
+            reset_coactivations=True,
         )
         new_defaults = {k: v for k, v in new_defaults.items() if k not in self.__dict__}
         self.__dict__.update(new_defaults)
 
+        # Set some attributes to be a list corresponding to self.dynamic_sparse_modules.
+        for attr in ["hebbian_prune_perc", "weight_prune_perc"]:
+            self._make_attr_iterable(attr, counterpart=self.sparse_modules)
+
+        for idx, (hprune, wprune) in enumerate(
+            zip(self.hebbian_prune_perc, self.weight_prune_perc)
+        ):
+            module = self.sparse_modules[idx]
+            module.hebbian_prune = hprune
+            module.weight_prune = wprune
+
         # initialize hebbian learning
-        self._init_coactivation_tracking()
+        self._init_hebbian()
         self.prune_cycles_completed = 0
 
-    def _init_coactivation_tracking(self):
-        """
-        Override method in children classes
-        Should only track coactivations if required by the algorithm
-        """
-        self.network.apply(init_coactivation_tracking)
+    def _init_hebbian(self):
+        for module in self.sparse_modules:
+            if self._is_dynamic(module):
+                module.init_coactivation_tracking()
+        if hasattr(self.network, "forward_with_coactivations"):
+            self.network.forward = self.network.forward_with_coactivations
+
+    def _is_dynamic(self):
```

Comment:
You have `_is_dynamic` take either 1 input (`self`) or two inputs (`self` and `module`), but you should keep its arguments consistent.

Reply:
Thanks. Fixed, will send another commit.